### PR TITLE
retrace: Refactor the RetraceWorker

### PR DESCRIPTION
**Pull request by mkutlak**
This PR replaces parts of the RetraceWorker code that deal with:
- creating worker threads
- keeping track of ones that are alive
- moving tasks from queue to queue

with just three lines of code.

1) Create a ThreadPoolExecutor
2) Submit tasks to the executor
3) Add a callback for submitted tasks

Signed-off-by: Martin Kutlak <mkutlak@redhat.com>

### CODE REVIEW

File: src/pyfaf/retrace.py [hunk 1]
```diff
@@ -98,62 +100,88 @@ def __init__(self, db_debug_package, db_src_package, bin_pkg_map, db=None):
...
-    def __init__(self, worker_id, inqueue, outqueue):
-        name = "Worker #{0}".format(worker_id)
-        super(RetraceWorker, self).__init__(name=name)
-        self.inqueue = inqueue
-        self.outqueue = outqueue
-        self.stop = False
-        # Instance of 'RootLogger' has no 'getChildLogger' member
+    def __init__(self, db, tasks, problemplugin, workers):
+
+        self.name = "RetraceWorker"
```
Comment (ernestask) [inline src/pyfaf/retrace.py]: RetracePool?

**Comment by mkutlak** [comment]
If you want to test this PR (with will-crash):

1. Download package for retracing
```
dnf install --downloadonly --destdir /var/tmp/local_repo --installroot /var/tmp/local_repo --releasever 30 abrt-addon-ccpp gdb rpm shadow-utils will-crash
```
```
dnf debuginfo-install --downloadonly --destdir /var/tmp/local_repo --installroot /var/tmp/local_repo --releasever 30 abrt-addon-ccpp gdb rpm shadow-utils will-crash
```

2. Createrepo
```
createrepo_c /var/tmp/local_repo
```

3. Pull-components
```
sudo -u faf faf pull-components -o faf --opsys-release 30
```

4. Add repo to faf
```
sudo -u faf faf repoadd f30 rpmmetadata file:///var/tmp/local_repo
```

5. Assign the repo to release
```
sudo -u faf faf repoassign f30 "Fedora 30" x86_64
```

6. will_crash

7. Save reports and create problems
```
sudo -u faf faf save-reports
sudo -u faf faf create-problems
```

8. Retrace core
```
sudo -u faf faf retrace -p core
```

File: src/pyfaf/retrace.py [hunk 2]
```diff
@@ -1,5 +1,6 @@
 import re
-import threading
+import collections
```
Comment (ernestask) [inline src/pyfaf/retrace.py]: Unused import.

File: src/pyfaf/retrace.py (hunk 1 above)
Comment (mkutlak) [reply src/pyfaf/retrace.py]: Ahh, but then the logs would be `faf.blabla.RetracePool: mumbo jumbo`.

Maybe change the format of the logs to show thread names instead?

Comment (ernestask) [reply src/pyfaf/retrace.py]: Yeah, that might be better.

Comment (mkutlak) [reply src/pyfaf/retrace.py]: I've changed it to `RetracePool` for now. 

I have some changes for logging that I would like solve in different PR.

File: src/pyfaf/retrace.py [hunk 3]
```diff
@@ -98,62 +99,88 @@ def __init__(self, db_debug_package, db_src_package, bin_pkg_map, db=None):
...
         """
-        Asynchronously unpack one set of packages (debuginfo, source, binary)
+        taskid = 0
+
+        with futures.ThreadPoolExecutor(max_workers=self.workers, thread_name_prefix=self.name) as executor:
+            while self.tasks:
+                taskid += 1
+                task = self.tasks.popleft()
+                try:
+                    future = executor.submit(self._process_task, task, taskid, self.db)
```
Comment (ernestask) [inline src/pyfaf/retrace.py]: `_process_task()` does not take a database parameter.

**Comment by ernestask** [comment]
Apart from one exception related to argument counts, seems to work fine.

File: src/pyfaf/retrace.py (hunk 3 above)
Comment (mkutlak) [reply src/pyfaf/retrace.py]: :man_facepalming:

**Pull request merged**